code/modules/power/singularity/emitter.dm

```diff
@@ -47,189 +70,213 @@
 
 
-/obj/machinery/power/emitter/emp_act(var/severity)//Emitters are hardened but still might have issues
-//	add_load(1000)
-/*	if((severity == 1)&&prob(1)&&prob(1))
-		if(src.active)
-			src.active = 0
-			src.use_power = 1	*/
-	return 1
+/obj/machinery/power/emitter/emp_act()
+	activate(null)
+	return TRUE
 
 /obj/machinery/power/emitter/machinery_process()
 	if(stat & (BROKEN))
 		return
-	if(src.state != 2 || (!powernet && active_power_usage))
-		src.active = 0
+	if(state != EMITTER_WELDED || (!powernet && active_power_usage))
+		active = FALSE
 		update_icon()
 		return
-	if(((src.last_shot + src.fire_delay) <= world.time) && (src.active == 1))
-
+	if(((last_shot + fire_delay) <= world.time) && active)
 		var/actual_load = draw_power(active_power_usage)
 		if(actual_load >= active_power_usage) //does the laser have enough power to shoot?
 			if(!powered)
-				powered = 1
+				powered = TRUE
 				update_icon()
-				investigate_log("regained power and turned <font color='green'>on</font>","singulo")
+				if(special_emitter)
+					investigate_log("regained power and turned <font color='green'>on</font>","singulo")
 		else
 			if(powered)
-				powered = 0
+				powered = FALSE
 				update_icon()
-				investigate_log("lost power and turned <font color='red'>off</font>","singulo")
+				if(special_emitter)
+					investigate_log("lost power and turned <font color='red'>off</font>","singulo")
 			return
 
-		src.last_shot = world.time
-		if(src.shot_number < burst_shots)
-			src.fire_delay = 2
-			src.shot_number ++
+		last_shot = world.time
+		if(shot_number < burst_shots)
+			fire_delay = 2
+			shot_number++
 		else
-			src.fire_delay = rand(min_burst_delay, max_burst_delay)
-			src.shot_number = 0
+			fire_delay = rand(min_burst_delay, max_burst_delay)
+			shot_number = 0
 
 		//need to calculate the power per shot as the emitter doesn't fire continuously.
-		var/burst_time = (min_burst_delay + max_burst_delay)/2 + 2*(burst_shots-1)
-		var/power_per_shot = active_power_usage * (burst_time/10) / burst_shots
+		var/burst_time = (min_burst_delay + max_burst_delay) / 2 + 2 * (burst_shots - 1)
+		var/power_per_shot = active_power_usage * (burst_time / 10) / burst_shots
 
-		playsound(src.loc, 'sound/weapons/emitter.ogg', 25, 1, 3, 0.5, 1)
+		playsound(get_turf(src), 'sound/weapons/emitter.ogg', 25, TRUE, 3, 0.5, TRUE)
 		if(prob(35))
 			spark_system.queue()
 
-		var/obj/item/projectile/beam/emitter/A = new /obj/item/projectile/beam/emitter( src.loc )
-		A.damage = round(power_per_shot/EMITTER_DAMAGE_POWER_TRANSFER)
+		var/obj/item/projectile/beam/emitter/A = new /obj/item/projectile/beam/emitter(get_turf(src))
+		A.damage = round(power_per_shot / EMITTER_DAMAGE_POWER_TRANSFER)
 		A.launch_projectile(get_step(src, dir))
 
 /obj/machinery/power/emitter/attackby(obj/item/W, mob/user)
-
+	if(istype(W, /obj/item/device/assembly/signaler))
+		var/obj/item/device/assembly/signaler/S = W
+		user.drop_from_inventory(W, src)
+		signaler = S
+		S.machine = src
```

Comment:
This seems a bit hardcoded and a pain to do for every machine. I'd personally handle this in machinery/attackby and maybe add a `CAN_HOLD_SIGNALER` flag?

Reply:
nah, not many machines return the parent proc when hit, and probably shouldn't, cuz that could cause things to get real fucky real fast